src/MobileRedirectManager.php

```diff
@@ -0,0 +1,305 @@
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
+		if ( ! AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) ) {
+			return false;
+		}
+		return self::should_redirect_via_js() || self::is_mobile_request();
+	}
+
+	/**
+	 * Determine if mobile redirection is available for the current request.
+	 *
+	 * @return bool True if available, false otherwise.
+	 */
+	public static function is_available_for_request() {
+		return self::is_enabled() && is_amp_available();
+	}
+
+	/**
+	 * Determine if the current request is from a mobile device.
+	 *
+	 * @return bool True if current request is from a mobile device, otherwise false.
+	 */
+	public static function is_mobile_request() {
+		/**
+		 * Filters whether the current request is from a mobile device. This is provided as a means to short-circuit
+		 * the normal determination of a mobile request below.
+		 *
+		 * @since 1.6
+		 *
+		 * @param null $is_mobile Whether the current request is from a mobile device.
+		 */
+		$pre_is_mobile = apply_filters( 'amp_pre_is_mobile', null );
+
+		if ( null !== $pre_is_mobile ) {
+			return (bool) $pre_is_mobile;
+		}
+
+		$current_user_agent = wp_unslash( $_SERVER['HTTP_USER_AGENT'] );
+
+		if ( empty( $current_user_agent ) ) {
+			return false;
+		}
+
+		$user_agents_regex = implode(
+			'|',
+			array_map(
+				static function ( $user_agent ) {
+					return preg_quote( $user_agent, '/' );
+				},
+				self::get_user_agents()
+			)
+		);
+
+		if ( preg_match( "/$user_agents_regex/", $current_user_agent ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Determine if mobile redirection should be done via JavaScript.
+	 *
+	 * @return bool True if mobile redirection should be done, false otherwise.
+	 */
+	public static function should_redirect_via_js() {
+		/**
+		 * Filters whether mobile redirection should be done via JavaScript. If false, a server-side solution will be used instead.
+		 *
+		 * @since 1.6
+		 *
+		 * @param bool $should_redirect_via_js Whether JS redirection should be used.
+		 */
+		return (bool) apply_filters( 'amp_redirect_via_js', true );
+	}
+
+	/**
+	 * Get a list of user agents to use for comparison against the user agent from the current request.
+	 *
+	 * @return string[] An array of user agents.
+	 */
+	public static function get_user_agents() {
+		// Default list compiled from the user agents listed in `wp_is_mobile()`.
+		$default_user_agents = [
+			'Mobile',
+			'Android',
+			'Silk/',
+			'Kindle',
+			'BlackBerry',
+			'Opera Mini',
+			'Opera Mobi',
+		];
+
+		/**
+		 * Filters the list of user agents used to determine if the user agent from the current request is a mobile one.
+		 *
+		 * @since 1.6
+		 *
+		 * @param string[] $user_agents List of user agents.
+		 */
+		return apply_filters( 'amp_mobile_user_agents', $default_user_agents );
```

Comment:
The method claims it will only every return an array, so it should enforce that, filters are not to be trusted.
```suggestion
		return (array) apply_filters( 'amp_mobile_user_agents', $default_user_agents );
```